10 add evaluation pipeline

J-Dymond opened the pull request:
# Evaluation pipeline

Has a few utils files containing metrics and utility functions, and some scripts which perform evaluation on a selected model. Will briefly go over the evaluation scripts as they are at the moment, and the changes I made to the evaluation dataset class. The evaluation scripts can be run periodically throughout training to allow get a clearer picture of model performance as it is being trained. 

### `quantitative_eval.py`
Performs quantitative evaluation over a test set with a selected model, it compares ground truth inputs against *perturbed* inputs like in the paper. Unlike the paper we don't generate these, rather they are answers to different questions, randomly sampled from within the same author. In the future we can change this according to our work package. The script outputs the truth ratio values, and raw losses which are outputted as a numpy array for further processing. Currently if running as __main__ these are saved to a `.np` file in a separate folder in the parent folder of that where the model weights are stored.

### `qualitative_eval.py`
This performs a qualitative evaluation of the model. Loops over the test data and generates an output answer according to the input question. These are both printed along with the target printed along with the target to allow qualitative comparison against the target answer. 

### `EvalQADataset()` Changes
I made some changes to allow the quantitative evaluation script to work. Namely adding a batch formatter which when given a question, outputs input IDs, labels, and attention masks with appropriate padding for batch computation. Furthermore, a method which locates *perturbed* answers is added, which when given a question index will locate a random question pertaining the same author which can be used as a perturbed answer.

J-Dymond committed:
added some preliminary evaluation metrics with some tests

J-Dymond committed:
added KS-Test to evaluation functions

J-Dymond committed:
standardising pytest.approx() in tests

J-Dymond committed:
corrected the input for conditional probability: it accepts normalised losses across potential answers

J-Dymond committed:
adding the loss function

J-Dymond committed:
added a loss and a test for it

J-Dymond committed:
used the dummy model + tokeniser for tests

J-Dymond committed:
TODO: write tests and debug eval functions

J-Dymond committed:
added tests for the truth ratio, now working as intented

J-Dymond committed:
adding evaluation pipeline

J-Dymond committed:
trying to implement the evaluation pieplline

J-Dymond committed:
added an end to end test in test_eval.py

J-Dymond committed:
added an end to end test in test_eval.py

J-Dymond committed:
added a qualitative evalation function

J-Dymond committed:
added argparser for the qualitative eval script

J-Dymond committed:
added initial quanititative and qualitative evaluation scripts

J-Dymond committed:
commented and provided docstrings for all code

J-Dymond committed:
added the data split to the input arguments of the evaluation script (defaults to forget)

jack89roberts committed:
:pushpin: Update poetry.lock

jack89roberts committed:
:fire: delete VSCode config

jack89roberts commented on src/arcsf/data/tofu.py:
```diff
@@ -154,4 +154,10 @@ def load_tofu(
         all_data[forget_indices]
     ), Dataset.from_dict(all_data[retain_indices])
 
+    setattr(forget_set, "TOFU_NUM_AUTHORS", TOFU_NUM_AUTHORS)
+    setattr(forget_set, "TOFU_Q_PER_AUTHOR", TOFU_Q_PER_AUTHOR)
+
+    setattr(retain_set, "TOFU_NUM_AUTHORS", TOFU_NUM_AUTHORS)
+    setattr(retain_set, "TOFU_Q_PER_AUTHOR", TOFU_Q_PER_AUTHOR)
```
These can be: `forget_set.TOFU_NUM_AUTHORS = TOFU_NUM_AUTHORS`, and slight preference to rename to `forget_set.num_authors = TOFU_NUM_AUTHORS` so the names used in the rest of the code are less TOFU specific. Also, I've not looked at the rest of the code yet but are these the right values to be using? E.g. the forget/retain sets may have fewer authors/questions than the full dataset. If the values do need to be whole-dataset values then maybe make that clear in the names (which could mean ignoring the renaming suggestion above and keeping TOFU_...).

J-Dymond replied:
added this change

jack89roberts commented on src/arcsf/data/data_module.py:
```diff
@@ -84,9 +100,12 @@ def __init__(
         self,
         data: Dataset,
```
Not a change in this PR so sorry I didn't pick up before, but is this actually meant to be a HuggingFace `Dataset` (`datasets.Dataset`) rather than a `torch` dataset (and throughout the code in typehints, docstrings etc.)? Unless we're explicitly making a torch dataset somewhere we're likely always working with HuggingFace datasets as that's what `load_dataset` returns.

J-Dymond replied:
I tried this and it broke a lot of tests, maybe we can have a look at this in a co-working session?

J-Dymond replied:
clarified type-hinting for these

jack89roberts commented on src/arcsf/data/data_module.py:
```diff
@@ -107,31 +129,119 @@ def __init__(
+        inp, _ = qa
+        author_q_n = row % self.data.TOFU_Q_PER_AUTHOR
+        perturbed_options = self.data[
+            row - author_q_n : row + self.data.TOFU_Q_PER_AUTHOR - author_q_n
+        ]["answer"]
```
My understanding of this via an example:

```
row = 45
TOFU_Q_PER_AUTHOR = 20
author_q_n = 45 % 20 = 5
self.data[ 45 - 5 : 45 + 20 - 5] = self.data[40 : 60]
```
-> So is this trying to get all the answers for the relevant author? (then deleting the original answer on the next line)

But, as per previous comment, is `self.data` the whole dataset here or a (forget/retain) subset? If a subset will computing the indices this way work as expected? Could you use the `author_index` / `quetion_index` columns added in `load_tofu` instead, for example?

J-Dymond replied:
Exactly, that's what its doing. But I recognise now this doesn't generalise to a shuffled set. As you say, it would be much better to use the author index to find the relevant items. I will make this change.

J-Dymond replied:
I think I've addressed this now

J-Dymond replied:
Though, my new implementation runs self.data.filter() everytime __getitem__ is called. I will think of a way to calculate these at initialisation and store them?

jack89roberts commented on src/arcsf/data/data_module.py:
```diff
@@ -107,31 +129,119 @@ def __init__(
+            row - author_q_n : row + self.data.TOFU_Q_PER_AUTHOR - author_q_n
+        ]["answer"]
+        del perturbed_options[author_q_n]
+        perturbed_options = perturbed_options[: self.n_perturbed]
```
Should this be a random sample?

J-Dymond replied:
We could randomly sample, this was a placeholder that I forgot to develop further. Perhaps I could randomly shuffle the perturbed_options array to keep this simple?

J-Dymond replied:
I've added a random shuffle now when its sampled, with the seed passed to the dataset

jack89roberts commented on src/arcsf/data/data_module.py:
```diff
@@ -107,31 +129,119 @@ def __init__(
+        ]["answer"]
+        del perturbed_options[author_q_n]
+        perturbed_options = perturbed_options[: self.n_perturbed]
+        formatted = [None] * (self.n_perturbed + 1)
+        formatted[0] = self.batch_formatter(qa)
+        for p_idx, perturbed in enumerate(perturbed_options):
+            formatted[p_idx + 1] = self.batch_formatter((inp, perturbed))
+        return formatted
```
```suggestion
        formatted = [self.batch_formatter((inp, perturbed)) for perturbed in perturbed_options]
        return [self.batch_formatter(qa)] + formatted
```

J-Dymond replied:
applied, thank you!

jack89roberts commented on src/arcsf/data/data_module.py:
```diff
@@ -107,31 +129,119 @@ def __init__(
         else:
             self.answer_sampler = self.get_answer
 
+        if qualitative_eval:
+            self.format = self.qualitative_formatter
```
Placeholder for further thought/discussion - would we like it to be possible to do both qualitative and quantative from a single EvalQADataset instance?

J-Dymond replied:
Yeah I could see that, I was keen to have as little logic as possible in the forward method to make it more efficient.. But I understand the bottleneck is generally the model inference. 

Maybe we could have an attribute which we can change to dictate the __getitem__ method? ie dataset.return_qualitative == True and dataset.return_quantitative == False, then when the __getitem__ method will look at these to work out what to return?

J-Dymond replied:
the dataset now has the option of doing both

J-Dymond replied:
I implemented something like written above. Setting this up for batch_computation could be investigated in the future.

jack89roberts commented on src/arcsf/data/data_module.py:
```diff
@@ -107,31 +129,119 @@ def __init__(
+        )
+        return (encoded_inp, encoded_tar)
+
+    def batch_formatter(
```
maybe easier to discuss separately but in the HuggingFace trainer the batching is usually handled by a separate data collator - there's an example on the forgetting branch. So this maybe could be replaced with a function from elsewhere.

J-Dymond replied:
Perhaps we can discuss this in a co-working session?

J-Dymond replied:
Discussed in co-working session. Can implement this when it is needed.

jack89roberts commented on src/arcsf/data/data_module.py:
```diff
@@ -107,31 +129,119 @@ def __init__(
+        tar = self.answer_sampler(idx)
 
-        return self.tokenizer(input), self.tokenizer(target)
+        return self.format((inp, tar), idx)
```
Placeholder whilst I look at the code: Is it necessary to return `idx` here?

J-Dymond replied:
it returns the output of self.format(), index is sometimes used in this function to identify the perturbed options

jack89roberts commented on src/arcsf/eval/metrics.py:
```diff
@@ -0,0 +1,39 @@
+import torch
```
Some comments/type hints/docstrings/references back to the TOFU paper would be nice here to help understand what's going on - mainly for truth ratio.

J-Dymond replied:
will do

jack89roberts commented on src/arcsf/eval/qualitative_eval.py:
```diff
@@ -0,0 +1,109 @@
+        n_inputs : number of inputs to sample
+        random_seed : Random seed for the dataloader
+        generate_kwargs : keyword arguments for the model.generate() method
+
+    Returns:
+        None
```
```suggestion
```

Not needed as it doesn't add anything over the `-> None` type hint.

J-Dymond replied:
done thank you

jack89roberts commented on src/arcsf/eval/qualitative_eval.py:
```diff
@@ -0,0 +1,109 @@
+    **generate_kwargs: dict,
+) -> None:
+    """Performs qualitative evaluation of the selected model over selected data.
+    Prints the generated text give a question followed by the ground truth target.
```
```suggestion
    Prints the generated text given a question followed by the ground truth target.
```

J-Dymond replied:
yes, this was a typo thank you

jack89roberts commented on src/arcsf/eval/qualitative_eval.py:
```diff
@@ -0,0 +1,109 @@
+            " against target strings."
+        )
+    )
+    parser.add_argument("directory", type=str, help="Relative path to model directory.")
```
`directory` -> `model_path`?

J-Dymond replied:
yes

jack89roberts commented on src/arcsf/eval/qualitative_eval.py:
```diff
@@ -0,0 +1,109 @@
+    tokenizer = GPT2Tokenizer.from_pretrained(model_dir)
+    model.config.pad_token_id = tokenizer.eos_token_id
+
+    experiment_config = yaml.safe_load(open(model_dir + "/experiment_config.yaml"))
+    forget_data, retain_data = get_data(
+        "tofu", random_seed=rand, **experiment_config["data_config"]
+    )
```
If using the experiment config the path should be configurable (and it probably could be possible to get the model path from there too). Or could take a data config only (also with a configurable path).

J-Dymond replied:
Yes, I agree. I can make this change to the argparser

jack89roberts commented on src/arcsf/eval/quantitative_eval.py:
```diff
@@ -0,0 +1,126 @@
+        batch_end_index = batch_idx * batch_size + len(gt_batch["input_ids"])
+        # don't need gradient
+        with torch.no_grad():
+            gt_dummy_model_output = model(**gt_batch)
```
`gt_dummy_model_output` -> `gt_outputs`?

J-Dymond replied:
Yes

jack89roberts commented on src/arcsf/eval/quantitative_eval.py:
```diff
@@ -0,0 +1,126 @@
+        for perturbed_index in range(1, n_perturbed + 1):
+            p_batch = batch[perturbed_index]
+            with torch.no_grad():
+                p_dummy_model_output = model(**p_batch)
```
`p_dummy_model_output` -> `p_outputs`?

J-Dymond replied:
yes

jack89roberts commented on src/arcsf/eval/quantitative_eval.py:
```diff
@@ -0,0 +1,126 @@
+
+def quantitative_eval(
+    model: torch.nn.Module,
+    dataset: Dataset,
```
is this expected to be an `EvalQADataset`?

J-Dymond replied:
Yes

jack89roberts commented on src/arcsf/eval/quantitative_eval.py:
```diff
@@ -0,0 +1,126 @@
+    print(f"Mean losses:\n\tground_truth:{means[0]}\n\tperturbed:{means[1:]}")
+    tr = truth_ratio(all_losses)
+
+    return tr, all_losses
```
Does/should this return all the metrics (e.g. rouge etc.)?

J-Dymond replied:
Can update this to also return the probability, the `all_losses` object should allow us to calculate this elsewhere, but will implement here too. I will also investigate where Rouge score should be returned.

J-Dymond replied:
Revisiting the paper, the probability is used when there are multiple choice answers. That is, the real author/world fact questions. So here we're not using them. Though, perhaps this is something that needs addressing?

J-Dymond replied:
Rouge score will be better placed in qualitative eval, since the inputs are strings. Though maybe it would be better to get all of the metrics from a single script? I have issues here though, formatting the inputs when using batches.

jack89roberts commented on tests/test_data_module.py:
```diff
@@ -107,10 +107,7 @@ def test_idk_targets(data):
-        _identity,
-        _identity,
-        loss_type="idk",
+        data, _identity, _identity, loss_type="idk", qualitative_eval=True
```
It seems a bit odd to me that this requires `qualitative_eval` to be `True` to work

J-Dymond replied:
otherwise it returns batch formatted inputs, qualitative evaluation outputs a (question,answer) tuple

J-Dymond replied:
By passing in the _identity function for the tokenizer and formatter it will just return the raw strings. Meaning we can directly compare them with the file they're sampled from. It just saves us having to decode the dataset output.

jack89roberts commented on tests/test_eval.py:
```diff
@@ -0,0 +1,168 @@
+    test_outputs_incorrect = test_outputs_correct * -1
+
+    correct = eval_accuracy(test_outputs_correct, test_targets)
+    assert correct["eval_accuracy"] == pytest.approx(1.0)
+
+    incorrect = eval_accuracy(test_outputs_incorrect, test_targets)
+    assert incorrect["eval_accuracy"] == pytest.approx(0.0)
```
These should be exact (try to only use `pytest.approx` where it's definitely needed):

```suggestion
    assert correct["eval_accuracy"] == 1.0

    incorrect = eval_accuracy(test_outputs_incorrect, test_targets)
    assert incorrect["eval_accuracy"] == 0.0
```

And a test of an intermediate value (between 0 and 1) would be a nice to have.

J-Dymond replied:
added an intermediate test as well

jack89roberts commented on tests/test_eval.py:
```diff
@@ -0,0 +1,168 @@
+    assert incorrect["eval_accuracy"] == pytest.approx(0.0)
+
+
+def test_conditional_probability():
```
Same comments as accuracy test apply here

jack89roberts commented on tests/test_eval.py:
```diff
@@ -0,0 +1,168 @@
+    assert torch.mean(eval_prob["conditional_probs"][0]).item() == pytest.approx(0.0)
+
+
+def test_ks_test():
```
test of a non-zero value would be good too

J-Dymond replied:
Done

jack89roberts commented on tests/test_eval.py:
```diff
@@ -0,0 +1,168 @@
+
+    loss = get_loss(test_output.logits, labels)
+
+    assert isinstance(loss, torch.Tensor)
+    assert loss.grad_fn
```
is this the right test for this? I don't think we care about gradients for the eval?

J-Dymond replied:
Good point, I've removed it

J-Dymond committed:
added a third evaluation script which outputs all scores

J-Dymond committed:
broken test: fixed, to do with device when running non-tokenized versions

J-Dymond committed:
added a script which aggregates the evaluation metrics

jack89roberts commented:
- Ensure we understand what's going on in the eval (e.g. document shapes etc.)
  - including NaNs appearing in places 
- Compute and return `alternative=greater` in KS test (as well as default).
- Try to understand why IDK model comes out worse for forgetting

jack89roberts commented:
It would be nice if the tests capture some of what we were trying to think through earlier today, e.g. check truth ratio of one of the dummy forget models is larger than the dummy fine-tuned model and similar. I haven't checked back through the tests so it might be that you've already done that.

J-Dymond committed:
added function which evaluates the model against a baseline model and outputs the metrics we want to track. Also hidden away/cleaned up  evaluation scripts and moved most functions to the utils file

J-Dymond committed:
minor error in tests going from type(var) == dtype --> isinstance(var, dtype)

J-Dymond committed:
added a comment in the ks_test

J-Dymond commented:
I've made some changes to my pull request now, I've added a function in `evaluate_model.py` called `evaluate_model` :

```
def evaluate_model(
    model: torch.nn.Module,
    base_truth_ratios_path: str,
    tokenizer: transformers.AutoTokenizer,
    experiment_config: dict,
) -> dict[float, float, float, float, float]:
```

It outputs a dictionary containing:

```
result_dict = {
        "mean_tr_retain": retain_tr,
        "mean_rouge_score": rouge_score,
        "forget_quality_1": forget_quality_one_sided,
        "forget_quality_2": forget_quality_two_sided,
        "model_utility": model_utilty,
    }
```
This should be everything we want to track in the wandb. I've added some tests testing it, and I've moved the old scripts I wrote into a `/scripts` folder within `eval`. The functions for these scripts are all contained in utils now. So within the `eval` folder there are just three files:

- `metrics.py` : containing the functions for the metrics
- `utils.py` : containing all of the functions used in the scripts folder and the evaluate_model function
- `evaluate_model.py` : containing the function for evaluating model, in hindsight, maybe this can be moved to utils.. but I leave that up to what you guys think is best

J-Dymond committed:
minor changes: fixed some imports and ensured the  function performs the max() function in table 1 of the tofu paper (it wasn't)

J-Dymond commented:
Just something minor I didn't explicitly point out in the above: the path for the base model truth ratios should currently be the relative path to where the **forget** truth ratios are stored. 

The all_eval script will calculate and save these, provided you give it the forget dataset. 

These are the only values that need to be stored locally for evaluate_model to run, everything else should be calculated within the function.

J-Dymond committed:
minor fixes: added max generated tokens, and gave evaluate_model.py a script if ran as main

J-Dymond committed:
added base values path to the argparser of the evaluate script

J-Dymond committed:
fixed bug in test: evaluate model should be passed kwargs for the model.generate() method

jack89roberts commented on src/arcsf/data/data_module.py:
```diff
@@ -69,11 +70,27 @@ def qa_formatter_basic(qa: tuple[str, str]) -> str:
 
 
-class EvalQADataset(Dataset):
+def qa_formatter_autoregression(qa: tuple[str, str, int]) -> str:
```
Maybe `qa_formatter_blank` or similar (i.e. nothing added other than question and answer).

J-Dymond replied:
Made this change

jack89roberts commented on src/arcsf/data/data_module.py:
```diff
@@ -57,7 +58,7 @@ def get_data(
 
 
-def qa_formatter_basic(qa: tuple[str, str]) -> str:
+def qa_formatter_basic(qa: tuple[str, str, int]) -> str:
```
Note to self: Forgetting branch has refactored QA formatters.

jack89roberts commented on src/arcsf/data/data_module.py:
```diff
@@ -107,34 +132,144 @@ def __init__(
+        return (encoded_inp, encoded_tar)
+
+    def batch_formatter(
+        self,
```
Note to self: Data collators/padding choices.

jack89roberts commented on src/arcsf/data/data_module.py:
```diff
@@ -107,34 +132,144 @@ def __init__(
+            perturbed_options = self.data.filter(
+                lambda sample: sample["author_index"] == author_n
+                and sample["question_index"] != question_n
+            ).shuffle(seed=self.rand_gen.seed())
```
I've not gone through to check but is setting the seed like this here ok or will it be giving the same perturbed samples every time? i.e. is this function used multiple times during an evaluation run and should/does it give different perturbed rows each time?

J-Dymond replied:
It will give the same question indices for each question. We can remove the seed if this isn't the functionality we want?

jack89roberts replied:
@J-Dymond - I'm still unsure about this, but in two ways:

- Looking at the docs of [torch.Generator.seed](https://pytorch.org/docs/stable/generated/torch.Generator.html#torch.Generator.seed) this will not produce a reproducible value (i.e. the fact the generator is initialised with a seed of 42 has no bearing on what `rand_gen.seed()` will return). I think the correct thing to do would be to use the [`generator` arg of `shuffle`](https://huggingface.co/docs/datasets/v2.20.0/en/package_reference/main_classes#datasets.Dataset.shuffle), but note this is only compatible with `numpy` generators not `torch` generators.

- I'm not sure we want the same perturbed answers to be returned every time this is called with the same `author_n` and `question_n` (within a script), but I'm not sure this is a big issue either.

Some of this may not matter once we have a dataset with actual perturbed answers.

-

J-Dymond replied:
Ok, I misunderstood this, sorry! I've added another attribute storing the random_seed. But I agree, we don't need this with the new dataset, and I'm not sure we want it either.

jack89roberts commented on src/arcsf/data/tofu.py:
```diff
@@ -154,4 +154,10 @@ def load_tofu(
         all_data[forget_indices]
     ), Dataset.from_dict(all_data[retain_indices])
 
+    forget_set.num_authors = TOFU_NUM_AUTHORS
+    forget_set.q_per_author = TOFU_Q_PER_AUTHOR
+
+    retain_set.num_authors = TOFU_NUM_AUTHORS
+    retain_set.q_per_author = TOFU_Q_PER_AUTHOR
```
Are these still needed/used?

J-Dymond replied:
I don't think so, will remove.

J-Dymond replied:
This was originally used when we sampled questions, but now we use the author index instead. So this isn't used any more.

jack89roberts commented on src/arcsf/eval/scripts/all_eval.py:
```diff
@@ -0,0 +1,81 @@
+
+if __name__ == "__main__":
+    # currently this is our random seed
+    rand = 42
```
if leaving a seed fixed here we should make an issue about it - I think it's only needed due to the way we're mimicing paraphrases/perturbed answers currently, so maybe it can be added to an issue about that.

J-Dymond replied:
I have made it an input argument

jack89roberts commented on src/arcsf/eval/scripts/all_eval.py:
```diff
@@ -0,0 +1,81 @@
+        )
+    )
+    batch_size = 1
+    parser.add_argument("directory", type=str, help="Relative path to model directory.")
```
`directory` -> `model_path`, `model_dir` or similar

jack89roberts commented on src/arcsf/eval/scripts/analysis.py:
```diff
@@ -0,0 +1,100 @@
+    verbose: bool = False,
+    plotting: bool = False,
+) -> dict[str, float, float, float, float]:
+    """
```
Type hinting for dicts is just key type then value type (not one for each value)


```suggestion
) -> dict[str, float]:
    """
```

jack89roberts commented on src/arcsf/eval/scripts/analysis.py:
```diff
@@ -0,0 +1,100 @@
+
+
+if __name__ == "__main__":
+    base_dir = "temp/20240507-233700-957103"
+
+    all_results = {}
+
+    dirs = os.listdir("temp")
+    for dir in dirs:
+        try:
+            results = run_analysis(base_dir, "temp/" + dir, plotting=True, verbose=True)
+        except FileNotFoundError:
+            continue
+        all_results[results.pop("experiment_name")] = results
+
+    with open("temp/results/all_results.json", "w") as f:
+        json.dump(all_results, f)
+
+    plt.savefig("temp/results/all_models_cdf.pdf")
```
all directories etc. here should be configuarable ideally

J-Dymond replied:
Ah yeah, sorry this was something I only intended to run locally to get our results.

J-Dymond replied:
I've ammended this now. However, there is no longer a script to aggregate them all into a single plot. I can make something new, or we can do this as and when we need it.

jack89roberts commented on src/arcsf/eval/scripts/qualitative_eval.py:
```diff
@@ -0,0 +1,77 @@
+    model_dir = args.model_path
+
+    # these are hardcoded for now
+    rand = 42
+    model = GPT2LMHeadModel.from_pretrained(model_dir)
+    tokenizer = GPT2Tokenizer.from_pretrained(model_dir)
+    model.config.pad_token_id = tokenizer.eos_token_id
```
if leaving hardcoded for this PR make an issue about fixing this later

J-Dymond replied:
Random seed was the only hardcoded value, it is now given as an input argument, and picks a random value if it is not specified.

jack89roberts commented on src/arcsf/eval/evaluate_model.py:
```diff
@@ -0,0 +1,124 @@
+    experiment_config: dict,
+    **generate_kwargs: dict,
+) -> dict[float, float, float, float, float]:
+    """
```
```suggestion
) -> dict[str, float]:
    """
```

jack89roberts commented on src/arcsf/eval/evaluate_model.py:
```diff
@@ -0,0 +1,124 @@
+        experiment_config: experiment config (currently a dictionary)
+
+    Returns:
+        dictionary : aggregated metrics for evaluation (generated using `get_metrics`)
```
```suggestion
        dictionary of aggregated metrics for evaluation (generated using `get_metrics`)
```

jack89roberts commented on src/arcsf/eval/evaluate_model.py:
```diff
@@ -0,0 +1,124 @@
+    device = get_device()
+
+    # create datasets
+    RetainDataset = EvalQADataset(
```
naming convention: `retain_dataset`

J-Dymond replied:
Done

jack89roberts commented on src/arcsf/eval/evaluate_model.py:
```diff
@@ -0,0 +1,124 @@
+        device=device,
+        random_seed=RAND,
+    )
+    ForgetDataset = EvalQADataset(
```
naming convention: `forget_dataset`

J-Dymond replied:
Done

jack89roberts commented on src/arcsf/eval/utils.py:
```diff
@@ -0,0 +1,299 @@
+    return xs, ys
+
+
+def combine_dicts(forget_dict, retain_dict):
```
could have a more informative name than `combine_dicts`

J-Dymond replied:
Addressed this now with a docstring

jack89roberts commented on src/arcsf/eval/metrics.py:
```diff
@@ -0,0 +1,98 @@
+
+def conditional_probability(
+    token_normalised_losses: torch.Tensor,
+) -> dict[torch.Tensor]:
```
```suggestion
) -> dict[str, torch.Tensor]:
```

but see next comment

J-Dymond replied:
addressed this

jack89roberts commented on src/arcsf/eval/metrics.py:
```diff
@@ -0,0 +1,98 @@
+    sum = torch.sum(probs, dim=-1)  # shape : n_samples
+    # transpose here ensures that the n_samples dimension is in the right place
+    cond_probs = probs.T / sum  # shape : n_perturbed x n_samples
+    return {"conditional_probs": cond_probs}
```
why not just return `cond_probs` here? The only place I can see that this function is used the values are immediately extracted from the dictionary again anyway.

J-Dymond replied:
I kept this since the original TOFU code did it, I wasn't sure why they used a dictionary either but I thought there might have been some logging tool using it or something. 

Anyway.. I can change it since it isn't needed.

jack89roberts commented on src/arcsf/eval/metrics.py:
```diff
@@ -0,0 +1,98 @@
+    return {"conditional_probs": cond_probs}
+
+
+def eval_rouge_recall(gen_output: str, ground_truth: str) -> dict[float, float]:
```
```suggestion
def eval_rouge_recall(gen_output: str, ground_truth: str) -> dict[str, float]:
```

jack89roberts commented on src/arcsf/eval/metrics.py:
```diff
@@ -0,0 +1,98 @@
+        ground_truth : ground truth string
+
+    Returns:
+        dict[float, float]: scores
```
```suggestion
        ROUGE scores
```

J-Dymond replied:
Thank you, I think I missed this auto fill from my docstring plugin

jack89roberts commented on src/arcsf/eval/metrics.py:
```diff
@@ -0,0 +1,98 @@
+    # getting conditional probability of the losses, this does perform a transpose
+    # which doesn't affect the output values, but also isn't strictly necessary.
+
+    # cond_probs shape: n_perturbed x n_samples
```
is this technically `n_perturbed + 1` as it includes the correct answer as well?

J-Dymond replied:
yes, this was a mistake, thank you for noticing

jack89roberts commented on src/arcsf/eval/utils.py:
```diff
@@ -0,0 +1,299 @@
+    Returns:
+        _description_
+    """
+    output_logits = output_logits[
+        ..., :-1, :
+    ].contiguous()  # shape: batch_size x (sequence_length-1) x vocab_size
+    shifted_labels = labels[
+        ..., 1:
+    ].contiguous()  # shape : batch_size x (sequence_length - 1)
```
```suggestion
    # shape: batch_size x (sequence_length-1) x vocab_size
    output_logits = output_logits[..., :-1, :].contiguous()

    # shape : batch_size x (sequence_length - 1)
    shifted_labels = labels[..., 1:].contiguous()  
```

jack89roberts commented on src/arcsf/eval/utils.py:
```diff
@@ -0,0 +1,299 @@
+        ..., 1:
+    ].contiguous()  # shape : batch_size x (sequence_length - 1)
+    # output_logits.transpose(-1, -2) shape: batch_size x vocab x (sequence_length - 1)
+    loss = _loss_function(output_logits.transpose(-1, -2), shifted_labels).sum(
+        dim=-1
+    )  # loss shape: batch_size
```
```suggestion
    # loss shape: batch_size
    loss = _loss_function(output_logits.transpose(-1, -2), shifted_labels).sum(dim=-1)  
```

jack89roberts commented on src/arcsf/eval/utils.py:
```diff
@@ -0,0 +1,299 @@
+def get_metrics(
+    base_truth_ratios: torch.Tensor,
+    test_values: dict[torch.Tensor],
+) -> dict[float, float, float, float, float]:
+    """
```
```suggestion
) -> dict[str, float]:
    """
```

jack89roberts commented on src/arcsf/eval/utils.py:
```diff
@@ -0,0 +1,299 @@
+
+
+def all_eval(
+    model: torch.nn.Module,
```
All model typehints maybe should be `transformers.PreTrainedModel` (but could be they all work with `torch` modules anyway)

```suggestion
    model: transformers.PreTrainedModel,
```

jack89roberts commented on src/arcsf/eval/utils.py:
```diff
@@ -0,0 +1,299 @@
+    n_perturbed = dataset.n_perturbed
+    data_loader = DataLoader(dataset, batch_size=batch_size, shuffle=False)
+    output_dict = {
+        "all_losses": torch.zeros(
+            (dataset.__len__(), n_perturbed + 1), dtype=torch.float64
+        ),
+        "truth_ratios": torch.zeros(dataset.__len__()),
+        "rougeL_recall": torch.zeros(dataset.__len__()),
```
Use `len` rather than calling `__len__` directly


```suggestion
        "all_losses": torch.zeros(
            (len(dataset), n_perturbed + 1), dtype=torch.float64
        ),
        "truth_ratios": torch.zeros(len(dataset)),
        "rougeL_recall": torch.zeros(len(dataset)),
```

J-Dymond replied:
Made this change. Since I was using it three times I just created a variable. Is there a reason we prefer len() over calling __len__ ?

jack89roberts commented on src/arcsf/eval/utils.py:
```diff
@@ -0,0 +1,299 @@
+def get_analysis_values(
+    model_dir: str,
+) -> dict[np.ndarray, np.ndarray, np.ndarray, torch.Tensor, torch.Tensor]:
+    """
```
```suggestion
) -> dict[str, np.ndarray | torch.Tensor]:
```

J-Dymond replied:
I actually thought I set all of these to torch.Tensors. I'll have a look at that, but will add this change

jack89roberts commented on src/arcsf/eval/utils.py:
```diff
@@ -0,0 +1,299 @@
+    vals["forget_losses"] = np.loadtxt(model_dir + "/eval/forget/all_losses.txt")
+    vals["retain_losses"] = np.loadtxt(model_dir + "/eval/retain/all_losses.txt")
+    vals["rouge_scores"] = np.loadtxt(model_dir + "/eval/retain/rougeL_scores.txt")
+    # we re-calculate the truth ratio, since torch calculated many as NaNs
```
could we check for nans and log a warning somewhere if they appear?

J-Dymond replied:
I'm not sure how to log warnings, but will investigate

J-Dymond replied:
Added this now, it checks all of them at the end, and logs an error if there is.

jack89roberts reviewed (changes requested):
I haven't got my head around it fully yet, will look again next week. I left a few comments about things being hardcoded but not for everything - for this PR they can be kept hardcoded but if so we should make an issue with everything that's left outstanding/will need to be changed for future runs/experiments.

J-Dymond committed:
addressed comments in pull request

J-Dymond committed:
removed the hard-coded random seed from

jack89roberts commented:
I've rebased this on the forgetting pipeline branch - so merge the forget branch PR before this one. I'm going to spend some time this afternoon thinking about how to get the eval working during forgetting.

jack89roberts committed:
update lock file

jack89roberts committed:
fix test

jack89roberts committed:
add eos_token as QAFormatter arg

add BlankQAFormatter, and refactor associated tests

jack89roberts committed:
fix arg name in all_eval

jack89roberts committed:
update use of experiment config in all_eval

jack89roberts committed:
add missing rouge1_recall

jack89roberts committed:
gpu -> cuda in device str

jack89roberts committed:
update exp config in qualitative eval script

jack89roberts committed:
remove unused random seed

jack89roberts commented:
@J-Dymond (when you're back) - I'm struggling to understand the use of the quantitative/qualitative options in `EvalQADataset`. E.g. the `all_eval` script doesn't seem to do anything qualitative but it uses an `EvalQADataset` with qualitative set to true and things break if I switch it to false.

jack89roberts committed:
update use of exp config in evaluate_model

jack89roberts commented on src/arcsf/eval/evaluate_model.py:
```diff
@@ -0,0 +1,126 @@
+
+    # combine dictionaries and load the baseline model truth ratios
+    analysis_vals = combine_dicts(forget_values, retain_values)
+    base_truth_ratios = torch.tensor(np.loadtxt(base_truth_ratios_path))
```
I've not checked the paper as of writing this, but shouldn't this be using base truth ratios from the retain data and the forget data here (rather than just one set of base truth ratios that's used in both the retain and forget evals)?

J-Dymond replied:
The base model truth ratios are only used in the forget quality, so we only need them from the forget set. We could load from the retain set as well for completeness, but I don't think they'll be used our the current implementation.

jack89roberts committed:
update use of experiment config for get_data in evaluate_model

jack89roberts commented:
Couple of observations:

- `all_eval` is very slow - can it be made faster? Maybe one for a different PR/issue.
- `all_eval` when called from `evaluate_model` seems even slower (3-4x slower)? Part of this seems to be because the `all_eval` script sets `padding=False` in `EvalQADataset`, but `evaluate_model` leave it as the default of True. But it's still 2x slower after that.
  - This was comparing an `ascent` forgetting model with a `retain` model - the `ascent` model is slower than the `retain` model for some reason.

J-Dymond committed:
quick fix for random seed in EvalQADataset.get_perturbed() and moved input to device in all_eval()

J-Dymond committed:
starting batch computation change

J-Dymond committed:
added an eval data collator, and simplified the EvalQADataset class as a result

J-Dymond committed:
minor fix: padding value in labels should be the ignore index (-100), not the eos/padding token

J-Dymond committed:
refactoring evaluate collate function to inclide huggingface data collators

J-Dymond committed:
added padding side argument/functionality to the collate function

J-Dymond committed:
minor changes to padding parameters, and batch size

J-Dymond committed:
refactored and commented the datacollator for the eval script

J-Dymond committed:
removed padding_value in place for the tokenizer in the data collator: now padding value is assigned using the pad_token_id/eos_token_id of the tokenizer, then raises an error if both are none. Now tests all run correctly.

J-Dymond committed:
added the position_ids to the data_collator output, as left padding affects model behaviour in the absence of position_ids. Also added a test to ensure model outputs are the same regardless of padding, with a more stringent requirement on the resultant loss being equal.

J-Dymond committed:
added some comments and added a  argument to the  script

J-Dymond committed:
fixed tests

J-Dymond committed:
work in progress simple data collator, currently raising ValueError

J-Dymond committed:
fixed test, but position embeddings cause issue on test_eval_end_to_end

J-Dymond committed:
position_ids are now correctly generated

J-Dymond committed:
fixed ks_test test

J-Dymond committed:
small change in the formatting function, labels are now masked until the start token of the answer, including the formatting tokens between the question and answer

J-Dymond committed:
removed arguments

J-Dymond committed:
some simplifications to the dataset

J-Dymond committed:
eval_end_to_end test currently doesn't pass due to normalised loss/dummy_model not generating eos token quirk

jack89roberts committed:
add eos token to QA preprocessing when training dummy models

jack89roberts committed:
update q/a start/end finding & simplify EvalQADataset

jack89roberts committed:
trim extra padding from questions

jack89roberts committed:
Eval loader/collator tweaks/fixes

- Fix random generator not being used in eval dataset
- Move moving tensors to device to collator rather than dataset

jack89roberts committed:
use accelerate to handle moving to device

jack89roberts committed:
wip refactoring towards trainer integration

jack89roberts committed:
more refactoring, delete now unused files/functions

jack89roberts committed:
make evaluator class

jack89roberts committed:
log mean losses

jack89roberts committed:
log mean forget tr

jack89roberts committed:
add a normal trainer that can run forget quality/utility eval

jack89roberts committed:
remove idk loss as option for evaluation

jack89roberts committed:
update eval load/saving in train script

jack89roberts committed:
remove loss_type in evaluator creation

jack89roberts committed:
fix "None" vs None bug

jack89roberts committed:
update test_tofu experiment configs

jack89roberts committed:
set eval batch size in train script

jack89roberts committed:
set max_new_tokens for generate

jack89roberts committed:
log level info

jack89roberts committed:
missing .forget_truth_ratios

jack89roberts committed:
fix log at info

J-Dymond committed:
added the generation private key file to the gitignore

J-Dymond reviewed (commented):
This was a test, please ignore
